pkg/referrerstore/oras/authprovider/k8secret_authprovider.go

```diff
@@ -0,0 +1,203 @@
+	if err != nil {
+		return nil, err
+	}
+
+	if conf.ServiceAccountName == "" {
+		conf.ServiceAccountName = defaultName
+	}
+
+	// get name of namespace ratify is running in
+	ratifyNamespace := os.Getenv(ratifyNamespaceEnv)
+	if ratifyNamespace == "" {
+		return nil, fmt.Errorf("environment variable %s not set", ratifyNamespaceEnv)
+	}
+
+	var k8secrets []*core.Secret
+	ctx := context.Background()
+
+	// iterate through config secrets, resolve each secret, and store in map
+	for _, k8secret := range conf.Secrets {
+		// default value of secret is assumed to be ratify namespace
+		if k8secret.Namespace == "" {
+			k8secret.Namespace = ratifyNamespace
+		}
+
+		secret, err := clusterClientSet.CoreV1().Secrets(k8secret.Namespace).Get(ctx, k8secret.SecretName, meta.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		// only dockercfg or docker config json secret type allowed
+		if secret.Type == core.SecretTypeDockercfg || secret.Type == core.SecretTypeDockerConfigJson {
+			k8secrets = append(k8secrets, secret)
+		} else {
+			return nil, fmt.Errorf("secret with unsupported type %s provided in config", secret.Type)
+		}
+	}
+
+	// get the the service account for ratify
+	serviceAccount, err := clusterClientSet.CoreV1().ServiceAccounts(ratifyNamespace).Get(ctx, conf.ServiceAccountName, meta.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	// extract the imagePullSecrets linked to service account
+	for _, imagePullSecret := range serviceAccount.ImagePullSecrets {
+		secret, err := clusterClientSet.CoreV1().Secrets(ratifyNamespace).Get(ctx, imagePullSecret.Name, meta.GetOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		// only dockercfg or docker config json secret type allowed
+		if secret.Type == core.SecretTypeDockercfg || secret.Type == core.SecretTypeDockerConfigJson {
+			k8secrets = append(k8secrets, secret)
+		}
+	}
+
+	return &k8SecretAuthProvider{
+		secrets: k8secrets,
+	}, nil
+}
+
+// Enabled checks if secrets list is not nil or empty
+func (d *k8SecretAuthProvider) Enabled(ctx context.Context) bool {
+	if d.secrets == nil || len(d.secrets) <= 0 {
+		return false
+	}
+
+	return true
+}
+
+// Provide finds secret corresponding to artifact's registry host name, extracts
+// the authentication credentials from k8 secret, and returns AuthConfig
+func (d *k8SecretAuthProvider) Provide(ctx context.Context, artifact string) (AuthConfig, error) {
+	if !d.Enabled(ctx) {
+		return AuthConfig{}, fmt.Errorf("k8 secret provider not properly enabled")
+	}
+
+	hostName, err := GetRegistryHostName(artifact)
+	if err != nil {
+		return AuthConfig{}, err
+	}
+
+	for _, secretLoaded := range d.secrets {
+		if secretLoaded.Type == core.SecretTypeDockercfg {
+			// if secret is a legacy docker config type
+			dockercfg, exists := secretLoaded.Data[core.DockerConfigKey]
+			if !exists {
+				return AuthConfig{}, fmt.Errorf("could not extract auth configs from .dockercfg")
+			}
+
```

Comment:
the logic in if/else is similar except dockerconfigkey vs dockerconfigjson key type , wonder if we could extract to a method

Reply:
I was able to condense the function to remove redundant code